hudi-common/src/main/java/org/apache/hudi/common/table/cdc/CDCExtractor.java

```diff
@@ -0,0 +1,359 @@
+   */
+  private void initInstantAndCommitMetadatas() {
+    try {
+      List<String> requiredActions = Arrays.asList(COMMIT_ACTION, DELTA_COMMIT_ACTION, REPLACE_COMMIT_ACTION);
+      HoodieActiveTimeline activeTimeLine = metaClient.getActiveTimeline();
+      Map<HoodieInstant, HoodieCommitMetadata> result = activeTimeLine.getInstants()
+          .filter(instant ->
+              instant.isCompleted()
+                  && isInRange(instant.getTimestamp(), startInstant, endInstant)
+                  && requiredActions.contains(instant.getAction().toLowerCase(Locale.ROOT))
+          ).map(instant -> {
+            HoodieCommitMetadata commitMetadata;
+            try {
+              if (instant.getAction().equals(HoodieTimeline.REPLACE_COMMIT_ACTION)) {
+                commitMetadata = HoodieReplaceCommitMetadata.fromBytes(
+                    activeTimeLine.getInstantDetails(instant).get(), HoodieReplaceCommitMetadata.class);
+              } else {
+                commitMetadata = HoodieCommitMetadata.fromBytes(
+                    activeTimeLine.getInstantDetails(instant).get(), HoodieCommitMetadata.class);
+              }
+            } catch (IOException e) {
+              throw new HoodieIOException(e.getMessage());
+            }
+            return Pair.of(instant, commitMetadata);
+          }).filter(pair ->
+              maybeChangeData(pair.getRight().getOperationType())
+          ).collect(Collectors.toMap(Pair::getLeft, Pair::getRight));
+      this.commits = result;
+    } catch (Exception e) {
+      throw new HoodieIOException("Fail to get the commit metadata for CDC");
+    }
+  }
+
+  private Boolean maybeChangeData(WriteOperationType operation) {
+    return operation == WriteOperationType.INSERT
+        || operation == WriteOperationType.UPSERT
+        || operation == WriteOperationType.DELETE
+        || operation == WriteOperationType.BULK_INSERT
+        || operation == WriteOperationType.DELETE_PARTITION
+        || operation == WriteOperationType.INSERT_OVERWRITE
+        || operation == WriteOperationType.INSERT_OVERWRITE_TABLE
+        || operation == WriteOperationType.BOOTSTRAP;
+  }
+
+  /**
+   * Parse HoodieWriteStat, judge which type the file is, and what strategy should be used to parse CDC data.
+   * Then build a [[ChangeFileForSingleFileGroupAndCommit]] object.
+   */
+  private CDCFileSplit parseWriteStat(
+      HoodieFileGroupId fileGroupId,
+      HoodieInstant instant,
+      HoodieWriteStat writeStat,
+      WriteOperationType operation) {
+    Path basePath = metaClient.getBasePathV2();
+    FileSystem fs = metaClient.getFs().getFileSystem();
+
+    CDCFileSplit cdcFileSplit;
+    if (StringUtils.isNullOrEmpty(writeStat.getCdcPath())) {
+      // no cdc log files can be used directly. we reuse the existing data file to retrieve the change data.
+      String path = writeStat.getPath();
+      if (path.endsWith(HoodieFileFormat.PARQUET.getFileExtension())) {
```

Comment:
No extensible to other file types

Reply:
let me upgrade this to support other base file types.

Reply:
to support other base file types like orc in spark is blocked by https://issues.apache.org/jira/browse/HUDI-4496.